lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs

```diff
@@ -199,13 +243,98 @@ data SelectionError
     | SelectionOutputsError ErrPrepareOutputs
     deriving (Eq, Show)
 
+-- | Accounts for pre-existing inputs provided to coin selection.
+--
+accountForExistingInputs :: Functor m => ModifySelection m change
+accountForExistingInputs performSelectionFn constraints params =
+    fmap modifyResult <$> performSelectionFn
+        (modifyConstraints constraints)
+        (modifyParams params)
+  where
+    modifyConstraints :: SelectionConstraints -> SelectionConstraints
+    modifyConstraints
+        = over #computeMinimumCost
+            modifyComputeMinimumCost
+        . over #computeSelectionLimit
+            modifyComputeSelectionLimit
+      where
+        modifyComputeMinimumCost
+            :: (SelectionSkeleton -> Coin)
+            -> (SelectionSkeleton -> Coin)
+        modifyComputeMinimumCost = (. modifySkeleton)
+          where
+            modifySkeleton :: SelectionSkeleton -> SelectionSkeleton
+            modifySkeleton =
+                over #skeletonInputCount (+ UTxO.size existingInputs)
+
+        modifyComputeSelectionLimit
+            :: ([TxOut] -> SelectionLimit)
+            -> ([TxOut] -> SelectionLimit)
+        modifyComputeSelectionLimit = (modifySelectionLimit .)
+          where
+            modifySelectionLimit :: SelectionLimit -> SelectionLimit
+            modifySelectionLimit = fmap $ subtract $ UTxO.size existingInputs
+
+    modifyParams :: SelectionParams -> SelectionParams
+    modifyParams
+        = over #assetsToMint
+            modifyAssetsToMint
+        . over #existingInputs
+            modifyExistingInputs
+        . over #rewardWithdrawal
+            modifyRewardWithdrawal
+        . over #utxoAvailable
+            modifyUTxOAvailable
+      where
+        modifyAssetsToMint :: TokenMap -> TokenMap
+        modifyAssetsToMint = TokenMap.add (view #tokens existingInputValue)
```

Comment:
Why are the existing input values added to assets to mint?

Reply:
Good question.

When we have pre-existing inputs, these are effectively "supplying extra value" to a coin selection.

The `assetsToMint` field currently provides the only well-tested way of supplying extra value.

By adding to this field here, we're effectively getting the ability to "add the value from several extra inputs" for free. We can also take advantage of the fact that the `assetsToMint` and `assetsToBurn` fields are already well covered with property tests. In a sense, we're treating the inner algorithm as a black box that we trust will do the right thing.

Of course, it would be nicer if the `assetsToMint` and `assetsToBurn` field names could reflect their generality. So we have a ticket for precisely that: ADP-1117